src/compiler/transformers/declarations.ts

```diff
@@ -265,9 +289,190 @@ export function transformDeclarations(context: TransformationContext) {
     let rawLibReferenceDirectives: readonly FileReference[];
     const resolver = context.getEmitResolver();
     const options = context.getCompilerOptions();
-    const { stripInternal } = options;
+    const { stripInternal, isolatedDeclarations } = options;
     return transformRoot;
 
+    function reportExpandoFunctionErrors(node: FunctionDeclaration | VariableDeclaration) {
+        resolver.getPropertiesOfContainerFunction(node).forEach(p => {
+            if (isExpandoPropertyDeclaration(p.valueDeclaration)) {
+                const errorTarget = isBinaryExpression(p.valueDeclaration) ?
+                    p.valueDeclaration.left :
+                    p.valueDeclaration;
+
+                context.addDiagnostic(createDiagnosticForNode(
+                    errorTarget,
+                    Diagnostics.Assigning_properties_to_functions_without_declaring_them_is_not_supported_with_isolatedDeclarations_Add_an_explicit_declaration_for_the_properties_assigned_to_this_function,
+                ));
+            }
+        });
+    }
+    function reportInferenceFallback(node: Node) {
+        if (!isolatedDeclarations || isSourceFileJS(currentSourceFile)) return;
+        if (isVariableDeclaration(node) && resolver.isExpandoFunctionDeclaration(node)) {
+            reportExpandoFunctionErrors(node);
+        }
+        else {
+            const heritageClause = findAncestor(node, isHeritageClause);
+            if (heritageClause) {
+                context.addDiagnostic(createDiagnosticForNode(node, Diagnostics.Extends_clause_can_t_contain_an_expression_with_isolatedDeclarations));
+            }
+            else {
+                context.addDiagnostic(getDiagnostic(node));
+            }
+        }
+
+        type WithSpecialDiagnostic =
+            | GetAccessorDeclaration
+            | SetAccessorDeclaration
+            | ShorthandPropertyAssignment
+            | SpreadAssignment
+            | ComputedPropertyName
+            | ArrayLiteralExpression
+            | SpreadElement
+            | FunctionDeclaration
+            | FunctionExpression
+            | ArrowFunction
+            | MethodDeclaration
+            | ConstructSignatureDeclaration
+            | BindingElement
+            | VariableDeclaration
+            | PropertyDeclaration
+            | ParameterDeclaration
+            | PropertyAssignment
+            | ClassExpression;
+        function getDiagnostic(node: Node) {

```

Comment:
Like the maps at the bottom of the file, this closure is probably best moved to `src\compiler\transformers\declarations\diagnostics.ts`, so all the error selection logic continues to live in the same place.

Reply:
Moved